com.unity.ml-agents.extensions/Runtime/Sensors/GridSensorBase.cs

```diff
@@ -0,0 +1,336 @@
+using Object = UnityEngine.Object;
+
+[assembly: InternalsVisibleTo("Unity.ML-Agents.Extensions.TestUtils")]
+namespace Unity.MLAgents.Extensions.Sensors
+{
+    /// <summary>
+    /// Grid-based sensor.
+    /// </summary>
+    public class GridSensorBase : ISensor, IBuiltInSensor, IDisposable
+    {
+        string m_Name;
+        Vector3 m_CellScale;
+        Vector3Int m_GridSize;
+        int[] m_ChannelDepths;
+        string[] m_DetectableObjects;
+        SensorCompressionType m_CompressionType;
+        ObservationSpec m_ObservationSpec;
+        internal BoxOverlapChecker m_BoxOverlapChecker;
+
+        // Buffers
+        float[] m_PerceptionBuffer;
+        Color[] m_PerceptionColors;
+        Texture2D m_PerceptionTexture;
+        float[] m_CellDataBuffer;
+
+        // Utility Constants Calculated on Init
+        int m_NumCells;
+        int m_CellObservationSize;
+        Vector3 m_CellCenterOffset;
+
+
+        public GridSensorBase(
+            string name,
+            Vector3 cellScale,
+            Vector3Int gridNum,
+            int[] channelDepths,
+            string[] detectableObjects,
+            SensorCompressionType compression
+        )
+        {
+            m_Name = name;
+            m_CellScale = cellScale;
+            m_GridSize = gridNum;
+            m_ChannelDepths = channelDepths;
+            m_DetectableObjects = detectableObjects;
+            m_CompressionType = compression;
+
+            if (m_GridSize.y != 1)
+            {
+                throw new UnityAgentsException("GridSensor only supports 2D grids.");
+            }
+
+            m_NumCells = m_GridSize.x * m_GridSize.z;
+            m_CellObservationSize = GetObservationSize();
+            m_ObservationSpec = ObservationSpec.Visual(m_GridSize.x, m_GridSize.z, m_CellObservationSize);
+            m_PerceptionTexture = new Texture2D(m_GridSize.x, m_GridSize.z, TextureFormat.RGB24, false);
+
+            ResetPerceptionBuffer();
+        }
+
+        public SensorCompressionType CompressionType
+        {
+            get { return m_CompressionType; }
+            set
+            {
+                if (!IsDataNormalized() && value == SensorCompressionType.PNG)
+                {
+                    Debug.LogWarning("Compression type {m_CompressionType} is only supported with normalized data. " +
+                        "The sensor will not compress the data.");
+                    return;
+                }
+                m_CompressionType = value;
+            }
+        }
+
+        internal float[] PerceptionBuffer
+        {
+            get { return m_PerceptionBuffer; }
+        }
+
+        protected string[] DetectableObjects
+        {
+            get { return m_DetectableObjects; }
+        }
+
+        /// <inheritdoc/>
+        public void Reset() { }
+
+        /// <summary>
+        /// Clears the perception buffer before loading in new data. If the gridDepthType is ChannelHot, then it initializes the
```

Comment:
I think this comment ("If the gridDepthType is ChannelHot...") is out of date now.